File: docx_to_txt/converters/numbering_converter.py
from docx_parsers.models.document_models import Paragraph
from docx_parsers.models.numbering_models import NumberingLevel


class NumberingConverter:
    """
    Class to convert numbering to plain text.
    """

    numbering_counters = {}

    @staticmethod
    def convert_numbering(paragraph: Paragraph, numbering_schema) -> str:
        """
        Convert numbering to plain text.
        :param paragraph: The paragraph object.
        :param numbering_schema: The numbering schema.
        :return: Plain text representation of the numbering.
        """
        numbering = paragraph.numbering
        level_key = (numbering.numId, numbering.ilvl)
        
        try:
            numbering_level = NumberingConverter.get_numbering_level(numbering_schema, numbering.numId, numbering.ilvl)
        except ValueError as e:
            print(f"Warning: {e}")
            return "•"

        if numbering.numId not in NumberingConverter.numbering_counters:
            NumberingConverter.numbering_counters[numbering.numId] = [0] * 9  # Supports up to 9 levels
        
        NumberingConverter.numbering_counters[numbering.numId][numbering.ilvl] += 1
        
        # Reset counters for deeper levels if a higher level is incremented
        for i in range(numbering.ilvl + 1, 9):
            NumberingConverter.numbering_counters[numbering.numId][i] = 0

        counters = NumberingConverter.numbering_counters[numbering.numId][:numbering.ilvl + 1]
        formatted_counters = [NumberingConverter.format_number(counters[i], numbering_level.numFmt) for i in range(numbering.ilvl + 1)]
        
        # Replace all placeholders in lvlText
        lvlText = numbering_level.lvlText
        for i in range(1, numbering.ilvl + 2):
            lvlText = lvlText.replace(f"%{i}", formatted_counters[i-1])

        return lvlText + " "

    @staticmethod
    def get_numbering_level(numbering_schema, numId: int, ilvl: int) -> NumberingLevel:
        instance = next((inst for inst in numbering_schema.instances if inst.numId == numId), None)
        if instance:
            level = next((lvl for lvl in instance.levels if lvl.ilvl == ilvl), None)
            if level:
                return level
        raise ValueError(f"Numbering level not found for numId: {numId}, ilvl: {ilvl}")

    @staticmethod
    def format_number(counter: int, numFmt: str) -> str:
        if numFmt == "decimal":
            return str(counter)
        elif numFmt == "lowerRoman":
            return NumberingConverter.to_roman(counter).lower()
        elif numFmt == "upperRoman":
            return NumberingConverter.to_roman(counter).upper()
        elif numFmt == "lowerLetter":
            return NumberingConverter.to_lower_letter(counter)
        elif numFmt == "upperLetter":
            return NumberingConverter.to_upper_letter(counter)
        elif numFmt == "bullet":
            return "•"
        return ""

    @staticmethod
    def to_roman(num: int) -> str:
        val = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
        syb = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]
        roman_num = ''
        i = 0
        while num > 0:
            for _ in range(num // val[i]):
                roman_num += syb[i]
                num -= val[i]
            i += 1
        return roman_num

    @staticmethod
    def to_upper_letter(num: int) -> str:
        return chr(64 + num)

    @staticmethod
    def to_lower_letter(num: int) -> str:
        return chr(96 + num)

File: docx_parser_converter_ts/src/docx_to_txt/converters/numberingConverter.ts
import type { Paragraph } from '../../docx_parsers/models/paragraphModels';
import type { NumberingSchema, NumberingLevel } from '../../docx_parsers/models/numberingModels';

/**
 * Converts DOCX numbering to plain text for TXT output, preserving the logic of the Python implementation.
 */
export class NumberingConverter {
    private static numberingCounters: Record<number, number[]> = {};

    public static resetCounters(): void {
        NumberingConverter.numberingCounters = {};
    }

    public static convertNumbering(paragraph: Paragraph, numberingSchema: NumberingSchema): string {
        const numbering = paragraph.numbering;
        if (!numbering) {
            // Python: broad try-except returns bullet if numbering is missing
            return '• ';
        }

        let numberingLevel: NumberingLevel;
        try {
            numberingLevel = NumberingConverter.getNumberingLevel(numberingSchema, numbering.numId, numbering.ilvl);
        } catch (e: any) {
            // Python: print warning and return bullet
            if (typeof console !== 'undefined') {
                console.warn(`Warning: ${e.message || e}`);
            }
            return '• ';
        }

        if (!NumberingConverter.numberingCounters[numbering.numId]) {
            NumberingConverter.numberingCounters[numbering.numId] = new Array(9).fill(0); // Up to 9 levels
        }
        NumberingConverter.numberingCounters[numbering.numId][numbering.ilvl]++;

        // Reset counters for deeper levels if a higher level is incremented
        for (let i = numbering.ilvl + 1; i < 9; i++) {
            NumberingConverter.numberingCounters[numbering.numId][i] = 0;
        }

        const counters = NumberingConverter.numberingCounters[numbering.numId].slice(0, numbering.ilvl + 1);
        const formattedCounters: string[] = [];
        for (let i = 0; i <= numbering.ilvl; i++) {
            formattedCounters.push(NumberingConverter.formatNumber(counters[i], numberingLevel.numFmt));
        }

        let lvlText = numberingLevel.lvlText;
        for (let i = 1; i <= numbering.ilvl + 1; i++) {
            const placeholder = `%${i}`;
            if (formattedCounters[i - 1] !== undefined) {
                lvlText = lvlText.replace(new RegExp(placeholder.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g'), formattedCounters[i - 1]);
            }
        }
        return lvlText + ' ';
    }

    private static getNumberingLevel(numberingSchema: NumberingSchema, numId: number, ilvl: number): NumberingLevel {
        const instance = numberingSchema.instances.find(inst => inst.numId === numId);
        if (instance) {
            const level = instance.levels.find(lvl => lvl.ilvl === ilvl);
            if (level) {
                return level;
            }
        }
        throw new Error(`Numbering level not found for numId: ${numId}, ilvl: ${ilvl}`);
    }

    private static formatNumber(counter: number, numFmt: string): string {
        switch (numFmt) {
            case 'decimal':
                return String(counter);
            case 'lowerRoman':
                return NumberingConverter.toRoman(counter).toLowerCase();
            case 'upperRoman':
                return NumberingConverter.toRoman(counter).toUpperCase();
            case 'lowerLetter':
                return NumberingConverter.toLowerLetter(counter);
            case 'upperLetter':
                return NumberingConverter.toUpperLetter(counter);
            case 'bullet':
                return '•';
            default:
                return '';
        }
    }

    private static toRoman(num: number): string {
        if (num <= 0) return String(num);
        const val = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
        const syb = ['M', 'CM', 'D', 'CD', 'C', 'XC', 'L', 'XL', 'X', 'IX', 'V', 'IV', 'I'];
        let romanNum = '';
        let currentNum = num;
        let i = 0;
        while (currentNum > 0 && i < val.length) {
            const count = Math.floor(currentNum / val[i]);
            for (let k = 0; k < count; k++) {
                romanNum += syb[i];
                currentNum -= val[i];
            }
            i++;
        }
        return romanNum;
    }

    private static toUpperLetter(num: number): string {
        return String.fromCharCode(64 + num);
    }

    private static toLowerLetter(num: number): string {
        return String.fromCharCode(96 + num);
    }
} 